### ISSUE TITLE
[amp-date-display] not discovered when inside script template

### ISSUE THREAD

**Issue by dritter** [ampproject/amphtml#24288]
Hi,

The validator does not recognize `amp-date-display` if this is used inside a `<script type="text/plain" template="amp-mustache">` tag. This does not happen on other components (like amp-carousel; see example).

![image](https://user-images.githubusercontent.com/1544760/63933988-9a884500-ca5a-11e9-8182-b18abbad3129.png)

Example:
```html
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <title>My AMP Page</title>
  <link rel="canonical" href="self.html" />
  <meta name="viewport" content="width=device-width,minimum-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <script async custom-element="amp-carousel" src="https://cdn.ampproject.org/v0/amp-carousel-0.2.js"></script>
  <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js"></script>
  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
  <script async custom-element="amp-date-display" src="https://cdn.ampproject.org/v0/amp-date-display-0.1.js"></script>
  <style amp-custom>
    h1 {
      margin: 1rem;
    }
  </style>
</head>
<body>
  <h1>Hello AMPHTML World!</h1>
  
  <amp-list src="https://cors-test.appspot.com/test"
            single-item
            items="."
            layout="fixed-height"
            height="50">
    
    <script type="text/plain" template="amp-mustache">
      <amp-date-display datetime="2006-02-14T15:27:33+01:00" layout="fixed-height" height="20">
      	<template type="amp-mustache">  	
          <small>{{ day }}.{{ month }}.{{ year }} {{ hourTwoDigit }}:{{ minuteTwoDigit }}</small>
		</template>
      </amp-date-display>
      <amp-carousel width="450"
        height="300">
        <amp-img src="/static/inline-examples/images/image1.jpg"
          width="450"
          height="300"></amp-img>
        <amp-img src="/static/inline-examples/images/image2.jpg"
          width="450"
          height="300"></amp-img>
        <amp-img src="/static/inline-examples/images/image3.jpg"
          width="450"
          height="300"></amp-img>
    </amp-carousel>
    </script>
  </amp-list>
</body>
</html>
```

This was first reported in #23731 

Edit:
The obvious workaround is to add an invisible `amp-date-display` outside the template..

**Comment by cathyxz** [comment]
@ampproject/wg-caching do any of you folks by chance know if there is some validator attribute we need to add to either `amp-mustache` or `amp-date-display` to make the validator recognize that `amp-date-display` is indeed used in the page in this case? It seems to work correctly for `amp-carousel`, so maybe we're just missing something in `amp-date-display`.

**Comment by honeybadgerdontcare** [comment]
This is due to the fact that some components/extensions are grandfathered into not requiring usage. So `amp-carousel` as a `script` tag does not require that it be used in the document, while newer components/extensions like `amp-date-display` do.

Note that you're using `<script type="text/plain" template="amp-mustache">` and then within that tag a `<template>` tag. Is that intentional? According to documentation one or the other is to be used....though I could be misunderstanding a detail here. @cathyxz 	is this some nuance with how `amp-date-display` works?

https://amp.dev/documentation/components/amp-mustache/

**Comment by Gregable** [comment]
This could even be in a manner that the validator wouldn't be able to detect, perhaps something like:

```
<{{amp-tag-name}}>
```

We could imagine that `<amp-list>` would be allowed to satisfy usage of all extension scripts. It's an easy way for publishers to "work around" the validator, but maybe it makes sense in this case?

Alternatively, could `amp-list` inject the required script tags if missing?

**Comment by cathyxz** [comment]
@honeybadgerdontcare you're right. The correct usage here would be to reference a `<template>` tag that's outside of the script tag. We don't allow template nesting. There was previously a bug where the template attribute wasn't allowed on `amp-date-display`, which was fixed in #23731. 

Is it the case that usage within `<template>` tags don't count as usage? I can potentially see a valid use case for components being only used within an amp-list and not the rest of the page, in which case they would fail validation.

**Comment by cathyxz** [comment]
> Alternatively, could amp-list inject the required script tags if missing?

It would add a lot of overhead for `amp-list` to parse and figure out which components need to be injected, and it's not possible for `amp-list` to know which versions of the components should be injected.

**Comment by honeybadgerdontcare** [comment]
> @honeybadgerdontcare you're right. The correct usage here would be to reference a `<template>` tag that's outside of the script tag. We don't allow template nesting. There was previously a bug where the template attribute wasn't allowed on `amp-date-display`, which was fixed in #23731.
> 
> Is it the case that usage within `<template>` tags don't count as usage? I can potentially see a valid use case for components being only used within an amp-list and not the rest of the page, in which case they would fail validation.
> 
> Oops--sorry, I see that @Gregable and I had a race condition in commenting here. Yes, that was the exact problem I was trying to describe.

The `template`  does account for usage. Putting the following in https://validator.ampproject.org is valid (removing the `script` tag and putting `template` around `amp-date-display`):

```
<!doctype html>
<html ⚡>
<head>
  <meta charset="utf-8">
  <title>My AMP Page</title>
  <link rel="canonical" href="self.html" />
  <meta name="viewport" content="width=device-width,minimum-scale=1">
  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
  <script async src="https://cdn.ampproject.org/v0.js"></script>
  <script async custom-element="amp-carousel" src="https://cdn.ampproject.org/v0/amp-carousel-0.2.js"></script>
  <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js"></script>
  <script async custom-element="amp-list" src="https://cdn.ampproject.org/v0/amp-list-0.1.js"></script>
  <script async custom-template="amp-mustache" src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"></script>
  <script async custom-element="amp-date-display" src="https://cdn.ampproject.org/v0/amp-date-display-0.1.js"></script>
  <style amp-custom>
    h1 {
      margin: 1rem;
    }
  </style>
</head>
<body>
  <h1>Hello AMPHTML World!</h1>
  
  <amp-list src="https://cors-test.appspot.com/test"
            single-item
            items="."
            layout="fixed-height"
            height="50">
    
  	<template type="amp-mustache">  	
      <amp-date-display datetime="2006-02-14T15:27:33+01:00" layout="fixed-height" height="20">
          <small>{{ day }}.{{ month }}.{{ year }} {{ hourTwoDigit }}:{{ minuteTwoDigit }}</small>
      </amp-date-display>
	</template>
      <amp-carousel width="450"
        height="300">
        <amp-img src="/static/inline-examples/images/image1.jpg"
          width="450"
          height="300"></amp-img>
        <amp-img src="/static/inline-examples/images/image2.jpg"
          width="450"
          height="300"></amp-img>
        <amp-img src="/static/inline-examples/images/image3.jpg"
          width="450"
          height="300"></amp-img>
    </amp-carousel>
  </amp-list>
</body>
</html>
```

**Comment by Gregable** [comment]
> Is it the case that usage within `<template>` tags don't count as usage? 

Yes. The contents of a `<script>` tag is just a text string until javascript runs. At least `<template>` is a DOM, just an inert one. The validator runs on the static dom and has no accurate way of knowing what the rendered result will be.

We could parse this specific string like we do for `<template>`, but then we're back to the problem that `<script>` was to solve, https://github.com/ampproject/amphtml/issues/11205.

It seems like in some cases, we want HTML parsing to run on the template itself, and in some cases we don't. It's very difficult for the code to distinguish the two cases.

**Comment by dritter** [comment]
@honeybadgerdontcare 

> Note that you're using `<script type="text/plain" template="amp-mustache">` and then within that tag a `<template>` tag. Is that intentional? According to documentation one or the other is to be used....though I could be misunderstanding a detail here. @cathyxz is this some nuance with how `amp-date-display` works?

By https://github.com/ampproject/amphtml/issues/21758#issuecomment-480999844 nested templates are allowed using this syntax (documented [here](https://amp.dev/documentation/components/amp-mustache/?format=websites#tables)). Anyway, referencing the template by id does not matter for this issue here.

@Gregable 
> It seems like in some cases, we want HTML parsing to run on the template itself, and in some cases we don't. It's very difficult for the code to distinguish the two cases.

This is especially hard for us, as the google search console runs its own validator that seems to operate on the rendered DOM, so we cannot tell if the produced code will be valid or not. But we are guilty of exploiting that hole in the past so that we can work around #23731. We already tried to switch to a proper `template` tag, but stumbled over the validation in the search console..

> We could imagine that <amp-list> would be allowed to satisfy usage of all extension scripts. It's an easy way for publishers to "work around" the validator, but maybe it makes sense in this case?

IMHO this is all caused by people trying to nest templates, and that case happens pretty quickly. Others seem to have similar problems (like #22924). We've done things like that (forms with templates inside `amp-list`) as well..

**Comment by Gregable** [comment]
It's probably also worth adding @choumx to this thread.

**Comment by choumx** [comment]
This may also be an issue in the future for `amp-script` where arbitrary AMP elements may be created at runtime via 3P JS.

How about adding a way for such elements to signal which extensions they require dynamically?

```html
<script type=text/plain template=amp-mustache 
  required-extensions="amp-date-display amp-foo">
...
</script>

<amp-script required-extensions="amp-date-display amp-bar">...</amp-script>
```

**Comment by cathyxz** [comment]
> How about adding a way for such elements to signal which extensions they require dynamically?

In this case, would `amp-mustache` be responsible for importing them (this would likely further delay rendering of templated elements) or would the validator need to be intelligent about parsing this and allowing imports used in `amp-mustache` only?

**Comment by choumx** [comment]
This would just be a signal to the validator to silence "extension not used" errors for `amp-date-display` and `amp-foo`.

**Comment by Gregable** [comment]
The obvious con to this is that you'd end up with folks adding to their website template:

```html
<script async custom-element='amp-foo'
  src='https://cdn.ampproject.org/v0/amp-foo-0.1.js'></script>
<script async custom-element='amp-bar'
  src='https://cdn.ampproject.org/v0/amp-bar-0.1.js'></script>
<script async custom-element='amp-baz'
  src='https://cdn.ampproject.org/v0/amp-baz-0.1.js'></script>

...
<!-- Make the AMP Validator happy so we don't have to trim required extensions
     per each document. -->
<script type=text/plain template=amp-mustache 
  required-extensions="amp-foo amp-bar amp-baz"></script>
</body></html>
```

Obviously there are other ways to to this today, but they are a little trickier. I'm not sure if this con is a good enough reason _not_ to do this though.

In general needing to statically determine ahead of time the set of dynamically necessary libraries is difficult, both for the publisher and amp.

**Comment by Gregable** [comment]
Could the main AMP runtime (not mustache) consume these `required-extensions` and upgrade them to `<script>` tags dynamically at the moment the document is loaded?

In this way, they can't even consume browser bandwidth until the rendering requirements are out of the way, but they can still begin fetch earlier than the loading of amp-mustache.

**Comment by choumx** [comment]
That's possible, but one issue is that could increase initial render latency for `amp-list` elements that are in first viewport (e.g. requires JS execution, no browser preload).

**Comment by Gregable** [comment]
I'm going to close this one as "working as intended" for validation. It's obviously not resolved satisfactorily, but I don't really see a path forward either.
